megatron/core/distributed/fsdp/mcore_fsdp_adapter.py

```diff
@@ -305,97 +298,44 @@ def sync_rng_states_across_tp_group(self):
 
 
 def _get_hsdp_tp_mesh(outer_fsdp_dp_group, dp_cp_group, tp_group):
-    assert HAVE_EINOPS, "einops is not installed. Please install it with `pip install einops`."
-    world_size = dist.get_world_size()
-
-    mesh = einops.rearrange(
-        torch.arange(world_size),
-        "(outer_fsdp_dp fsdp tp) -> outer_fsdp_dp fsdp tp",
-        outer_fsdp_dp=outer_fsdp_dp_group.size(),
-        tp=tp_group.size(),
-    )
-
-    mesh_fsdp_ranks = einops.rearrange(
-        mesh,
-        'outer_fsdp_dp fsdp tp -> (outer_fsdp_dp tp) fsdp',
-        tp=tp_group.size(),
-        fsdp=dp_cp_group.size(),
-    )
-    fsdp_group_ranks = dist.get_process_group_ranks(dp_cp_group)
-    assert _check_mesh_ranks_and_group_ranks_are_consistent(mesh_fsdp_ranks, fsdp_group_ranks), (
-        f"[Megatron-FSDP] FSDP ranks in the mesh {mesh_fsdp_ranks} "
-        f"do not match the ranks in the FSDP group {fsdp_group_ranks}."
-    )
-
-    mesh_tp_ranks = einops.rearrange(
-        mesh,
-        'outer_fsdp_dp fsdp tp -> (outer_fsdp_dp fsdp) tp',
-        tp=tp_group.size(),
-        fsdp=dp_cp_group.size(),
-    )
-    tp_group_ranks = dist.get_process_group_ranks(tp_group)
-    assert _check_mesh_ranks_and_group_ranks_are_consistent(mesh_tp_ranks, tp_group_ranks), (
-        f"[Megatron-FSDP] Tensor Parallel ranks in the mesh {mesh_tp_ranks} "
-        f"do not match the ranks in the TP group {tp_group_ranks}."
-    )
-
-    mesh_outer_fsdp_dp_ranks = einops.rearrange(
-        mesh,
-        'outer_fsdp_dp fsdp tp -> (fsdp tp) outer_fsdp_dp',
-        tp=tp_group.size(),
-        fsdp=dp_cp_group.size(),
-    )
-    outer_fsdp_dp_group_ranks = dist.get_process_group_ranks(outer_fsdp_dp_group)
-    assert _check_mesh_ranks_and_group_ranks_are_consistent(
-        mesh_outer_fsdp_dp_ranks, outer_fsdp_dp_group_ranks
-    ), (
-        f"[Megatron-FSDP] Outer FSDP Data Parallel ranks in the mesh {mesh_outer_fsdp_dp_ranks} "
-        f"do not match the ranks in the Outer FSDP DP group {outer_fsdp_dp_group_ranks}."
-    )
-
-    return mesh
-
+    if tp_group is None:
+        tp_ranks = [dist.get_rank()]
+    else:
+        tp_ranks = dist.get_process_group_ranks(tp_group)
+
+    if dp_cp_group is None:
+        dp_cp_tp_ranks = [tp_ranks]
+    else:
+        dp_size = dist.get_world_size(dp_cp_group)
+        dp_cp_tp_ranks = [None for _ in range(dp_size)]
+        dist.all_gather_object(dp_cp_tp_ranks, tp_ranks, group=dp_cp_group)
```

Comment:
Hmm, because we use all_gather_objects, we cannot make these two calls async... :(

Reply:
Do you mean it can be optimized into an all_gather operation?